plugins/woocommerce/src/StoreApi/Utilities/OrderController.php

```diff
@@ -556,14 +556,20 @@ public function validate_selected_shipping_methods( $needs_shipping, $chosen_shi
 			throw $exception;
 		}
 
-		$valid_methods = array_keys( WC()->shipping()->get_shipping_methods() );
+		$packages = WC()->shipping()->get_packages();
+		foreach ( $packages as $package_id => $package ) {
+			$chosen_rate_for_package    = wc_get_chosen_shipping_method_for_package( $package_id, $package );
+			$valid_rate_ids_for_package = array_map(
```

Comment:
Opted to go for this `array_map` and get the ID rather than just doing `array_keys` on `$package['rates']` because I'm not sure the array's keys are guaranteed to match the rate, are they? This is definitely safer and I am more confident shipping it.

We can add a followup to investigate if `array_keys` can be used instead.